src/dbnode/persist/fs/migration/migration.go

```diff
@@ -0,0 +1,101 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package migration
+
+import (
+	"github.com/m3db/m3/src/dbnode/namespace"
+	"github.com/m3db/m3/src/dbnode/persist"
+	"github.com/m3db/m3/src/dbnode/persist/fs"
+	"github.com/m3db/m3/src/dbnode/storage"
+	xtime "github.com/m3db/m3/src/x/time"
+)
+
+// Migration interface is implemented by tasks that wish to perform a data migration
+// on a fileset. Typically involves updating files in a fileset that were created by
+// a previous version of the database client
+type Migration interface {
+
+	// Run is the set of steps to successfully complete a migration
+	Run() error
+}
+
+var _ Migration = &ToVersion1_1{}
+
+// ToVersion1_1 is an object responsible for migrating a fileset to version 1.1
+type ToVersion1_1 struct {
+	newMergerFn fs.NewMergerFn
+	infoFile    fs.ReadInfoFileResult
+	shard       uint32
+	namespace   namespace.Metadata
+	opts        storage.Options
+}
+
+func NewToVersion1_1(infoFile fs.ReadInfoFileResult, shard uint32, namespace namespace.Metadata, opts storage.Options) Migration {
+	return ToVersion1_1{
+		newMergerFn: fs.NewMerger,
+		infoFile:    infoFile,
+		shard:       shard,
+		namespace:   namespace,
+		opts:        opts,
+	}
+}
+
+// Run executes the steps to bring a fileset to Version 1.1
+func (v ToVersion1_1) Run() error {
+	reader, err := fs.NewReader(v.opts.BytesPool(), v.opts.CommitLogOptions().FilesystemOptions())
+	if err != nil {
+		return err
+	}
+
+	merger := v.newMergerFn(reader, v.opts.DatabaseBlockOptions().DatabaseBlockAllocSize(),
+		v.opts.SegmentReaderPool(), v.opts.MultiReaderIteratorPool(),
+		v.opts.IdentifierPool(), v.opts.EncoderPool(), v.opts.ContextPool(), v.namespace.Options())
+
+	volIndex := v.infoFile.Info.VolumeIndex
+	fsID := fs.FileSetFileIdentifier{
+		Namespace:   v.namespace.ID(),
+		Shard:       v.shard,
+		BlockStart:  xtime.FromNanoseconds(v.infoFile.Info.BlockStart),
+		VolumeIndex: volIndex,
+	}
+
+	nsCtx := namespace.NewContextFrom(v.namespace)
```

Comment:
I've seen instances where this is created with a lock held (e.g. https://github.com/m3db/m3/blob/master/src/dbnode/storage/namespace.go#L1061) and places where it's not (e.g. https://github.com/m3db/m3/blob/master/src/dbnode/storage/bootstrap/bootstrapper/fs/source.go#L345). My assumption based on some quick code reading is this is safe to do during the bootstrap (which is when this migration will be executed). Is that accurate?

Reply:
I'm not sure either (didn't dig) but it seems like it? If we're not taking locks to grab ns ctx in the fs bootstrapper it should be fine here. Maybe just add some comments somewhere that fs migrations can only be run during bootstrap?

Also, to verify that this is true (it looks like we only really need the read lock since the namespace `Schema` can change) we could look at whether it's possible to receive ns schema updates during bootstrap.

Reply:
Hey this is fine when using "NewContextFrom(...)" because it calls "nsMetadata.Options().SchemaHistory().GetLatest()" which is synchronized.

In `namespace.go` it's used to get the latest schema without having to do further synchronization than necessary and a lot of methods already are holding the namespace read lock in `namespace.go` hence why it's slightly different in that file.